Windows path fixes

ammachado opened the pull request:
Fixes #958

fabapp2 commented:
HI @ammachado 
Thanks for picking this 🚀 
I started on `fix/958-fix-build-with-windows` but my Windows "machine" makes progressing hard (it's very slow) and I missed opening a draft PR.
Would you want to integrate my changes into your branch to bring our changes together?
We should try to unify paths as centralized as possible. E.g. In `RewriteSoruceFileHolder#getSourcePath()` and alike.

fabapp2 commented on components/sbm-core/src/test/java/org/springframework/sbm/project/resource/TestProjectContext.java:
```diff
@@ -575,18 +568,23 @@ private Integer getOrder(ProjectResourceWrapper l1) {
-                    replacedBean.put(ApplicationEventPublisher.class, eventPublisher);
+                // Test performance improvement on Windows
+                if (SystemUtils.IS_OS_WINDOWS) {
+                    replacedBeans.put(MavenPomCache.class, new InMemoryMavenPomCache());
```
Let's not dynamically change behavior in a test-harness.
The performance change largely depends on the state of the `MavenPomCache` (filesystem when using RocksDB). 
If it has the dependencies it's faster, otherwise it might even be even slower.

Nevertheless, it would be great to find a reliable way to avoid the need for downloads during test runs, but the cached poms have additional metadata and thus can't be taken from local Maven repo. Whereas the RocksDB approach sometimes gave trouble with locks when used in tests.

fabapp2 replied:
Wow, that's a lot. Did I get you right that the ÌnMemoryCache` dropped your build-time?

fabapp2 commented on components/sbm-core/src/test/java/org/springframework/sbm/project/resource/TestProjectContext.java:
```diff
@@ -575,18 +568,23 @@ private Integer getOrder(ProjectResourceWrapper l1) {
+                }
+
+                if (eventPublisher != null) {
+                    replacedBeans.put(ApplicationEventPublisher.class, eventPublisher);
```
This doesn't work as the bean implementing `ApplicationEventPublisher` is the `ApplicationContext` itself.
The whole "bean replacement" approach needs a rehaul.

fabapp2 commented:
Something was broken, when running `mvn clean install`, I get

````bash
[ERROR] Errors: 
[ERROR]   BootUpgrade_27_30_IntegrationTest.migrateSimpleApplication:45->IntegrationTestBaseClass.scanProject:211 » IllegalArgument Can't read recipes from resource 'spring-context-xml-import.yaml'
[ERROR]   BootUpgrade_27_30_MultiModule_IntegrationTest.migrateMultiModuleApplication:43->IntegrationTestBaseClass.scanProject:211 » IllegalArgument Can't read recipes from resource 'spring-context-xml-import.yaml'
[ERROR]   MigrateJpaApplicationIntegrationTest.migrateJpaApplication:35->IntegrationTestBaseClass.scanProject:211 » IllegalArgument Can't read recipes from resource 'spring-context-xml-import.yaml'
[ERROR]   MigrateJaxRsAnnotationsRecipeIntegrationTest.happyPath:136->IntegrationTestBaseClass.scanProject:211 » IllegalArgument Can't read recipes from resource 'spring-context-xml-import.yaml'
[ERROR]   MigrateJsf2xToSpringBootRecipeIntegrationTest.testRecipe:57->IntegrationTestBaseClass.scanProject:211 » IllegalArgument Can't read recipes from resource 'spring-context-xml-import.yaml'
[ERROR]   MigrateXmlBeanConfigurationToJavaConfigIntegrationTest.testRecipe:76->IntegrationTestBaseClass.scanProject:211 » IllegalArgument Can't read recipes from resource 'spring-context-xml-import.yaml'
[INFO] 
[ERROR] Tests run: 13, Failures: 0, Errors: 6, Skipped: 7
[INFO] 

````

These tests require a Docker container and thus don't run (yet) in GH actions.

fabapp2 commented on components/sbm-recipes-boot-upgrade/src/test/java/org/springframework/sbm/boot/upgrade_27_30/report/SpringBootUpgradeReportActionTest.java:
```diff
@@ -160,7 +160,7 @@ void verifyRenderedHtml(@TempDir Path tempDir) throws IOException {
                         <groupId>org.springframework.boot</groupId>
                         <artifactId>spring-boot-starter-parent</artifactId>
-                        <version>2.7.3</version>
+                        <version>2.7.2</version>
```
What made you change this?

ammachado replied:
I'm using a machine that hasn't direct access to Maven Central, and on the Artifactory instance I'm using, this artifact is flagged, so I downgraded it.

fabapp2 replied:
OK ;) 
It doesn't hurt, all tests still passing.

fabapp2 commented on components/sbm-recipes-spring-framework/src/main/java/org/springframework/sbm/actions/spring/xml/include/ImportSpringXmlConfigAction.java:
```diff
@@ -39,25 +39,28 @@
-    @Setter
-    @Autowired
     @JsonIgnore
-    private Configuration freemarkerConf;
```
This does not work. The Action is created by Jackson using no-args constructor.
Teh Spring beans get injected later, see [here](https://github.com/spring-projects-experimental/spring-boot-migrator/blob/7e4f2a1a0249eff24a42c348e9c9e3421242e195/components/sbm-core/src/main/java/org/springframework/sbm/engine/recipe/DefaultActionDeserializer.java#L28)

fabapp2 commented on components/sbm-recipes-spring-framework/src/test/java/org/springframework/sbm/actions/spring/xml/include/ImportSpringXmlConfigXmlToJavaConfigurationActionTest.java:
```diff
@@ -78,8 +78,7 @@ void oneXmlBeansFile() {
-        ImportSpringXmlConfigAction action = new ImportSpringXmlConfigAction();
-        action.setFreemarkerConf(configuration);
+        ImportSpringXmlConfigAction action = new ImportSpringXmlConfigAction(configuration);
         action.apply(ctx);
```
For prod this constructor can't be used, see the comment above.

fabapp2 commented:
@ammachado are you running under Windows? 
I have a hard time testing the build with the Windows machine I have at hand.

ammachado commented on components/sbm-core/src/main/java/org/springframework/sbm/build/migration/MavenPomCacheProvider.java:
```diff
@@ -15,36 +15,27 @@
-
-	private MavenPomCache inMemory() {
-		return new InMemoryMavenPomCache();
+	public MavenPomCacheProvider(Optional<MavenPomCache> mavenPomCacheProvider) {
```
I thought about using Spring's `ObjectProvider` instead.

fabapp2 replied:
Wouldn't an overloaded constructor do the job while being simple and concise?
BTW, The PomCache configuration feature is going to be replaced, see https://github.com/spring-projects-experimental/spring-boot-migrator/tree/version/revamp/sbm-support-rewrite#maven-pom-cache

fabapp2 commented on components/sbm-core/src/main/java/org/springframework/sbm/common/util/OsAgnosticPathMatcher.java:
```diff
@@ -33,8 +33,7 @@
 
-	private PathMatcher pathMatcher = new AntPathMatcher();
-	private LinuxWindowsPathUnifier pathUnifier = new LinuxWindowsPathUnifier();
+	private final PathMatcher pathMatcher = new AntPathMatcher();
```
I've seen the comment. It sounds right. Just wondering why no tests were failing.
I think replacing the AntPathMatcher with a `FileSystems.getDefault().getPathMatcher(pattern)` would surface a problem and make the code more streamlined.

fabapp2 replied:
The problem with windows as in #958 should be fixed with the usage of `OsAgnosticPahMatcher` now.

fabapp2 commented:
HI @ammachado thanks for your adjustments!
I only had a glimpse but didn't immediately see how the last commit relates to the Windows paths problem?
If it isn't, please create a new issue describing the requirements for the last commit, prepare a separate PR, and move this commit there.
I will not be available next week and would like to merge your changes and mine into `main` asap as these will then hopefully make `main` run under Windows again. Anything related can go into this PR.
Also, I had a few comments about the calls in some test, could you remove these calls as they're redundant?
Thank you!

ammachado commented:
> HI @ammachado thanks for your adjustments! I only had a glimpse but didn't immediately see how the last commit relates to the Windows paths problem? If it isn't, please create a new issue describing the requirements for the last commit, prepare a separate PR, and move this commit there. I will not be available next week and would like to merge your changes and mine into `main` asap as these will then hopefully make `main` run under Windows again. Anything related can go into this PR. Also, I had a few comments about the calls in some test, could you remove these calls as they're redundant? Thank you!

I removed the calls and moved the extra commit to a separate branch. I'm also away from my Windows machine until the second week of November.
